src/main/sessions.ts

```diff
@@ -420,12 +422,118 @@ class JupyterLabSession {
             this._window.show();
         });
 
-        this._window.loadURL(url.format({
-            pathname: path.join(__dirname, '../browser/index.html'),
-            protocol: 'file:',
-            slashes: true,
-            search: encodeURIComponent(JSON.stringify(this.info))
-        }));
+        const DESKTOP_APP_ASSETS_PATH = 'desktop-app-assets';
+
+        this._window.loadURL(`http://localhost:${appConfig.jlabPort}/${DESKTOP_APP_ASSETS_PATH}/index.html?${encodeURIComponent(JSON.stringify(this.info))}`);
+
+        const cookies: Map<string, string> = new Map();
+
+        const parseCookie = (cookie: string): Electron.CookiesSetDetails => {
+            const parts = cookie.split(';');
+            const nameVal = parts[0].split('=');
+            const name = nameVal[0].trim();
+            const value = nameVal[1].trim();
+
+            const cookieObj: Electron.CookiesSetDetails = {
+                url: `http://localhost:${appConfig.jlabPort}`,
+                name: name,
+                value: value
+            };
+
+            for (let i = 1; i < parts.length; ++i) {
+                const nameVal = parts[i].split('=');
+                const name = nameVal[0].trim();
+
+                switch (name) {
+                    case 'expires':
+                        {
+                            const value = nameVal[1].trim();
+                            cookieObj.expirationDate = Date.parse(value);
+                        }
+                        break;
+                    case 'Path':
+                        {
+                            const value = nameVal[1].trim();
+                            cookieObj.path = value;
+                        }
+                        break;
+                    case 'HttpOnly':
+                        cookieObj.httpOnly = true;
+                        break;
+                }
+            }
+
+            return cookieObj;
+        };
+
+        this._window.webContents.session.protocol.interceptBufferProtocol("http", (req, callback) => {
+            const desktopAppAssetsPrefix = `http://localhost:${appConfig.jlabPort}/${DESKTOP_APP_ASSETS_PATH}`;
+            const appAssetsDir = path.normalize(path.join(__dirname, '../../'));
+    
+            if (req.url.startsWith(desktopAppAssetsPrefix)) {
+                let assetPath = req.url.substring(desktopAppAssetsPrefix.length + 1);
+                const qMark = assetPath.indexOf('?');
+                if (qMark !== -1) {
+                    assetPath = assetPath.substring(0, qMark);
+                }
+                const assetFilePath = path.join(appAssetsDir, assetPath);
```

Comment:
What if assets path includes `..`? That could potentially escalate access to filesystem. I think that we need to be very stringent here and check that the new `assetFilePath` is a child of `appAssetsDir`.

Reply:
We could also hard-code a list of static files that the client is allowed to access.

Reply:
I am not sure if asset path can include `..`. we are calculating it from request url. but still, It makes sense to  check if assetFilePath is a child of appAssetsDir, good idea.